### PR TITLE
proxy penetration, moved metadata to meta structure

### PR DESCRIPTION
**Pull request by newtectonics**
New Contract and Token structures for flexible expansion

### CODE REVIEW

File: credmark/types/data/contract.py [hunk 1]
```diff
@@ -33,49 +40,126 @@ class Config:
         }
 
     def __init__(self, **data):
-        if isinstance(data.get('abi'), str):
```
Comment (newtectonics) [inline credmark/types/data/contract.py:36]: Did I break something

Comment (newtectonics) [reply credmark/types/data/contract.py:36]: I should put this back

File: credmark/types/data/contract.py [hunk 2]
```diff
@@ -33,49 +40,126 @@ class Config:
...
-                    raise ModelNoContextError(
-                        'No current context. Unable to create contract instance.')
-            else:
-                # This should never actually happen so we consider it a coding error
-                raise ModelRunError(
-                    'Contract address is None. Unable to create contract instance.')
-        return self._instance
-
-    def load(self):
+    def __load__(self):
```
Comment (newtectonics) [inline credmark/types/data/contract.py:52]: `__load not __load__`

File: credmark/types/data/contract.py [hunk 3]
```diff
@@ -33,49 +40,126 @@ class Config:
...
         context = credmark.model.ModelContext.current_context
         if context is not None:
             contract_q_results = context.run_model(
                 'contract.metadata', {'contractAddress': self.address})
```
Comment (newtectonics) [inline credmark/types/data/contract.py:56]: make this a core model

File: credmark/types/data/contract.py [hunk 4]
```diff
@@ -33,49 +40,126 @@ class Config:
...
-                # This should never actually happen so we consider it a coding error
-                raise ModelRunError(
-                    'Contract address is None. Unable to create contract instance.')
-        return self._instance
-
-    def load(self):
+    def __load__(self):
         context = credmark.model.ModelContext.current_context
         if context is not None:
             contract_q_results = context.run_model(
```
Comment (newtectonics) [inline credmark/types/data/contract.py:55]: change to new syntax

File: credmark/types/data/contract.py [hunk 5]
```diff
@@ -33,49 +40,126 @@ class Config:
...
 
+    @property
+    def events(self):
+        if self.proxy_for is not None:
+            return self.proxy_for.events
+        return self.instance.events
+
+    @property
+    def info(self):
+        if not self._loaded:
```
Comment (newtectonics) [inline credmark/types/data/contract.py:119]: `check in __load__ `

File: credmark/types/data/contract.py [hunk 6]
```diff
@@ -33,49 +40,126 @@ class Config:
...
                 res = contract_q_results['contracts'][0]
-                self.contract_name = res.get('contract_name')
-                self.constructor_args = res.get('constructor_args')
-                self.protocol = res.get('protocol')
-                self.product = res.get('product')
-                self.abi = res.get('abi')
+                self._meta.contract_name = res.get('contract_name')
+                self._meta.constructor_args = res.get('constructor_args')
+                self._meta.abi = res.get('abi')
+                if self.is_transparent_proxy:
```
Comment (newtectonics) [inline credmark/types/data/contract.py:63]: ugly logic on `self.proxy_for.address` lazy load

File: credmark/types/data/token.py [hunk 7]
```diff
@@ -36,110 +36,134 @@ def get_fungible_token_from_configuration(
...
-    is_native_token: bool = False
-    decimals: int
+
+    class TokenMetadata(Contract.ContractMetaData):
+        symbol: Union[str, None] = None
+        name: Union[str, None] = None
+        decimals: Union[int, None] = None
+        total_supply: Union[TokenWei, None] = None
+
+    _meta: TokenMetadata = PrivateAttr(default=TokenMetadata())
```
Comment (iwyrkore) [inline credmark/types/data/token.py:49]: Should be default_factory

File: credmark/types/data/token.py [hunk 8]
```diff
@@ -6,22 +6,22 @@
...
 from .data_content.fungible_token_data import FUNGIBLE_TOKEN_DATA
-from .data_content.erc_standard_data import ERC20_BASE_ABI
 from typing import List, Union
-from credmark.dto import PrivateAttr, IterableListGenericDTO, DTOField
-from ..models.core import CoreModels
+from credmark.dto import PrivateAttr, IterableListGenericDTO, DTOField, DTO
 
 
-def get_fungible_token_from_configuration(
+def get_token_from_configuration(
```
Comment (newtectonics) [inline credmark/types/data/token.py:13]: create token registry class

File: credmark/types/data/contract.py (hunk 1 above)
Comment (newtectonics) [reply credmark/types/data/contract.py:36]: fixed

File: credmark/types/data/contract.py (hunk 4 above)
Comment (newtectonics) [reply credmark/types/data/contract.py:59]: can't yet

File: credmark/types/data/contract.py (hunk 3 above)
Comment (newtectonics) [reply credmark/types/data/contract.py:60]: done

File: credmark/types/data/contract.py (hunk 5 above)
Comment (newtectonics) [reply credmark/types/data/contract.py]: fixed

**Pull request merged**